### ISSUE TITLE
Incomplete replays result in both GlobalGamePlayers having won=false

### ISSUE THREAD

**Issue by beheh** [HearthSim/HSReplay.net#110]
So we've got [this replay](https://hsreplay.net/admin/games/gamereplay/71486483/change/) (with this [GlobalGame](https://hsreplay.net/admin/games/globalgame/68762395/change/)), that got uploaded with an incomplete log where both players had the `PlayState.PLAYING` as their final playstate.

This resulted in the `GlobalGamePlayer.won` property being false (see GlobalGamePlayer), which is then pushed into the `GameReplay.won` NullBooleanField on upload (see `processing.py` with `find_or_create_replay`). This results in both `GlobalGamePlayers` having the field `won == false`, so users see the replay as a loss in their game result list. A user reported this and this is how we were made aware of this issue.

It should rather be `null`/`None`, to show that "we're not sure what it is".

## Proposal

```diff
diff --git a/hsreplaynet/games/models.py b/hsreplaynet/games/models.py
index 55f12c1..825d058 100644
--- a/hsreplaynet/games/models.py
+++ b/hsreplaynet/games/models.py
@@ -205,7 +205,12 @@ class GlobalGamePlayer(models.Model):
 
        @property
        def won(self):
-               return self.final_state in (PlayState.WINNING, PlayState.WON)
+               if self.final_state in (PlayState.WINNING, PlayState.WON):
+                       return True
+               if self.final_state in (PlayState.PLAYING):
+                       # we can't reliably tell
+                       return None
+               return False
 
        @property
        def opponent(self):
@@ -386,6 +391,8 @@ class GameReplay(models.Model):
                                state = "Disconnected"
                        elif self.won:
                                state = "Won"
+                       elif self.won is None:
+                               state = "Unknown"
                        elif friendly[1] == opponent[1]:
                                state = "Tied"
                        else:

```

**Comment by beheh** [comment]
The frontend can now handle and idsplay won=null from the API for unknown games, after the user reported the same issue with another game, but the backend still never sets this (as described above, it'll fall back to won=false). 

Can someone review this and land it? I don't want to push this without review, since this touches processing.

**Comment by beheh** [comment]
The reason this surfaces is because HSTracker is sending incomplete games and that's another issue by itself, but we should still be handling this correctly (otherwise we might as well drop the won=unknown state, if we're not going to use it).

**Comment by jleclanche** [comment]
Overloading the `won` property for this is just weird. It's fine to check for `final_state` in the second chunk you're changing.

**Comment by beheh** [comment]
Okay, then what about this:
```diff
diff --git a/hsreplaynet/games/models.py b/hsreplaynet/games/models.py
index ded8da3..0fc17df 100644
--- a/hsreplaynet/games/models.py
+++ b/hsreplaynet/games/models.py
@@ -229,6 +229,15 @@ class GlobalGamePlayer(models.Model):
        def won(self):
                return self.final_state in (PlayState.WINNING, PlayState.WON)
 
+       @property
+       def lost(self):
+               return self.final_state in (
+                       PlayState.LOSING,
+                       PlayState.LOST,
+                       PlayState.DISCONNECTED,
+                       PlayState.CONCEDED,
+               )
+
        @property
        def opponent(self):
                opponent_id = 2 if self.player_id == 1 else 2
diff --git a/hsreplaynet/games/processing.py b/hsreplaynet/games/processing.py
index 66b4cf7..78c9815 100644
--- a/hsreplaynet/games/processing.py
+++ b/hsreplaynet/games/processing.py
@@ -180,7 +180,7 @@ def find_or_create_replay(parser, entity_tree, meta, upload_event, global_game,
                "resumable": meta.get("resumable"),
                "build": meta["build"],
                "upload_token": upload_event.token,
-               "won": friendly_player.won,
+               "won": True if friendly_player.won else False if friendly_player.lost else None,
                "replay_xml": replay_xml_path,
                "hsreplay_version": hsreplay_version,
                "hslog_version": hslog_version,
```

**Comment by jleclanche** [comment]
No, your previous patch is fine if you change it like I mentioned. `won` is a boolean property which tells you whether the game was won. If it's `true`, you know the game was won. If it's `false`, you don't know anything, you just know that in that replay the player is not shown to be winning.

**Comment by beheh** [comment]
I think you're missing the point, I don't really care mind the boolean `won` property. I care about the API: We currently have a nullable `won` field in the API for Game Replays ([example API result](https://hsreplay.net/api/v1/games/vZEz7JoNnfgVo34HJUHhEF/)). It's specifically nullable [here](https://github.com/HearthSim/HSReplay.net/blob/master/hsreplaynet/games/models.py#L373) (as opposed to to the subsequent fields in that code). That field in the DB is currently populated during processing around the lines the second patch touches (in processing.py).

Since right now the property will always be a true boolean (never None), we do not take advantage of the nullable field in the DB, and our frontend (or any other API user) cannot differentiate between a "lost" GameReplay (won = false) and an incomplete log file (that might've even been won!). I'm not complaining about the definition or functionality of the won property. I do understand the purpose of that field, namely whether the player has exactly won (`PlayState.WINNING` or `WON`), but for the sake of the API I want some way for us to be able to say "we don't know how this game ended who won".

I therefore propose one of three solutions:
* We make use of the nullable field and use one of my patches, e.g. the first one, that changes the models definition of won returning None/null, or the second one, that adds the lost field and None/nulls it if neither is sure. I personally prefer the second patch.
* We don't make use of the nullable field (and can even go ahead and just make it non-nullable), and add another field to the GameReplay to express whether we know the result. For example we add a "lost" field similar to how the second patch works.
* We forget about the `won` field in the API and have the frontend and API users manually inspect the GlobalGamePlayer's final GameState to figure out if the game actually ended correctly (or if both Players ended in `GameState.PLAYING`). Obviously I could go ahead and do this right now, but this is logic we might need elsewhere, and any API user might need it as well.

Why do I think we need this? Because we've had users [complain](https://twitter.com/MarlonRichert/status/818168400413421568) about incomplete replays being marked as lost. Sure, this is a client bug, but there's no reason for use to not communicate this state to the frontend/API users, since we can easily do so, and I don't want to have to tell API users or resort myself to solution three.

**Comment by jleclanche** [comment]
I mentioned this on discord before: the property being nullable is not intended, it should be a plain boolean. Haven't gotten around to changing it yet but overloading that now is not a good idea.

 For more granularity, `final_state` is exposed. I don't see the problem.

**Comment by beheh** [comment]
Then we go with solution three. I'm changing the issue to make won non-nullable.
